### PR TITLE
Reify unbounded recursive types

### PR DESCRIPTION
**Pull request by lorenzleutgeb**
I ran into following issue with the new `reify` feature. Consider the added test case: `typeVariableMap` will look like:

```
"S" -> "java.lang.Enum<S>"   // where S from both sides are identical
```

This leads to an infinite recursion and in reality to stack overflow. What is the desired result in this case?

### CODE REVIEW

**Comment by jhalterman** [comment]
I think the best thing to do is see if we can return a result, such as a ParameterizedType, that mirrors the circular nature of the variable `S`.

**Comment by lorenzleutgeb** [comment]
Yes, that is what I came up with, too. And since that cycle might be of a length greater than one, we will need a cycle-check by means of keeping an explicit stack in `reify`. I'll implement that soon.

**Comment by lorenzleutgeb** [comment]
Implemented this through self-references in the returned instance. Please review.

File: src/main/java/net/jodah/typetools/TypeResolver.java [hunk 1]
```diff
@@ -140,6 +141,17 @@ public Type getOwnerType() {
       return original.getOwnerType();
     }
 
+    /*package-private*/ void setReifiedTypeArguments(Type[] reifiedTypeArguments) {
+      if (sealed)
+        throw new IllegalStateException("type arguments can only be set once");
```
Comment (jhalterman) [inline src/main/java/net/jodah/typetools/TypeResolver.java]: Is there a way to test hitting this condition?

File: src/main/java/net/jodah/typetools/TypeResolver.java [hunk 2]
```diff
@@ -479,33 +491,58 @@ public static Type resolveGenericType(Class<?> type, Type subType) {
...
+  private static Type reify(Type genericType, final Map<TypeVariable<?>, Type> typeVariableMap, Map<Type, Type> cache) {
+    // Terminal cases.
+    if (genericType == null)
       return null;
-    } else if (genericType instanceof Class<?>) {
+    else if (genericType instanceof Class<?>)
       return genericType;
-    } else if (genericType instanceof ParameterizedType) {
+    else if (cache.containsKey(genericType)) {
+      while (cache.containsKey(genericType)) {
```
Comment (jhalterman) [inline src/main/java/net/jodah/typetools/TypeResolver.java]: Can you describe when the while loop becomes necessary?

File: src/main/java/net/jodah/typetools/TypeResolver.java [hunk 3]
```diff
@@ -519,22 +556,27 @@ private static Type reify(final Type genericType, final Map<TypeVariable<?>, Typ
...
 
       final Type[] upperBounds = typeVariable.getBounds();
       // NOTE: According to https://docs.oracle.com/javase/tutorial/java/generics/bounded.html
       // if there are multiple upper bounds where one bound is a class, then this must be the
       // leftmost/first bound. Therefore we blindly take this one, hoping is the most relevant.
       // Hibernate does the same when erasing types, see also
       // https://github.com/hibernate/hibernate-validator/blob/6.0/engine/src/main/java/org/hibernate/validator/internal/util/TypeHelper.java#L181-L186
-      return reify(upperBounds[0], typeVariableMap);
+      cache = new HashMap<>(cache);
+      cache.put(typeVariable, upperBounds[0]);
```
Comment (jhalterman) [inline src/main/java/net/jodah/typetools/TypeResolver.java]: Could you describe a bit about why we create a new cache here, and also a few lines above?

File: src/main/java/net/jodah/typetools/TypeResolver.java [hunk 4]
```diff
@@ -479,33 +491,58 @@ public static Type resolveGenericType(Class<?> type, Type subType) {
...
   /**
    * Works like {@link #resolveRawClass(Type, Class, Class)} but does not stop at raw classes. Instead, traverses
    * referenced types.
    */
-  private static Type reify(final Type genericType, final Map<TypeVariable<?>, Type> typeVariableMap) {
-    if (genericType == null) {
+  private static Type reify(Type genericType, final Map<TypeVariable<?>, Type> typeVariableMap, Map<Type, Type> cache) {
+    // Terminal cases.
+    if (genericType == null)
       return null;
```
Comment (jhalterman) [inline src/main/java/net/jodah/typetools/TypeResolver.java]: Will this condition ever be hit?

File: src/main/java/net/jodah/typetools/TypeResolver.java [hunk 5]
```diff
@@ -479,33 +491,58 @@ public static Type resolveGenericType(Class<?> type, Type subType) {
...
+
   /**
    * Works like {@link #resolveRawClass(Type, Class, Class)} but does not stop at raw classes. Instead, traverses
    * referenced types.
    */
-  private static Type reify(final Type genericType, final Map<TypeVariable<?>, Type> typeVariableMap) {
-    if (genericType == null) {
+  private static Type reify(Type genericType, final Map<TypeVariable<?>, Type> typeVariableMap, Map<Type, Type> cache) {
+    // Terminal cases.
+    if (genericType == null)
```
Comment (jhalterman) [inline src/main/java/net/jodah/typetools/TypeResolver.java]: Will this condition ever be hit?

File: src/main/java/net/jodah/typetools/TypeResolver.java [hunk 6]
```diff
@@ -479,33 +491,58 @@ public static Type resolveGenericType(Class<?> type, Type subType) {
...
+      return reify(genericType, typeVariableTypeMap, new HashMap<>());
+  }
+
   /**
    * Works like {@link #resolveRawClass(Type, Class, Class)} but does not stop at raw classes. Instead, traverses
    * referenced types.
    */
-  private static Type reify(final Type genericType, final Map<TypeVariable<?>, Type> typeVariableMap) {
-    if (genericType == null) {
+  private static Type reify(Type genericType, final Map<TypeVariable<?>, Type> typeVariableMap, Map<Type, Type> cache) {
```
Comment (jhalterman) [inline src/main/java/net/jodah/typetools/TypeResolver.java]: We should probably add a javadoc comment mentioning what the `cache` is for.

**Comment by jhalterman** [comment]
Looks good overall. Mostly my comments were wanting to clarify some things for my own understanding :)

File: src/main/java/net/jodah/typetools/TypeResolver.java (hunk 6 above)
Comment (lorenzleutgeb) [reply src/main/java/net/jodah/typetools/TypeResolver.java]: Well, it is not exposed publicly and just there to find cycles anyway, so I didn't comment it, but will do.

File: src/main/java/net/jodah/typetools/TypeResolver.java (hunk 5 above)
Comment (lorenzleutgeb) [reply src/main/java/net/jodah/typetools/TypeResolver.java]: Good question... Will be hard to hit this from a public method. I will try to come up with an example. Pen and paper required...

File: src/main/java/net/jodah/typetools/TypeResolver.java (hunk 6 above)
Comment (jhalterman) [reply src/main/java/net/jodah/typetools/TypeResolver.java]: Sure - I was just thinking for future maintainers / readers of the code.

File: src/main/java/net/jodah/typetools/TypeResolver.java (hunk 5 above)
Comment (lorenzleutgeb) [reply src/main/java/net/jodah/typetools/TypeResolver.java]: No, this is never being hit. I removed the branch.

File: src/main/java/net/jodah/typetools/TypeResolver.java (hunk 2 above)
Comment (lorenzleutgeb) [reply src/main/java/net/jodah/typetools/TypeResolver.java]: For when the cycle inside the type structure is of a length greater than one, it must be traversed.

**Comment by lorenzleutgeb** [comment]
I improved these changes a bit and added some comments. I hope it makes the cut soon, we're distributing patched versions in our office...

**Comment by jhalterman** [comment]
Looks good 👍 Will cut a release shortly.

**Comment by lorenzleutgeb** [comment]
Thanks a lot! Waiting for that release :raised_hands:

**Comment by jhalterman** [comment]
Just pushed 0.6.1. Should be in central shortly.

**Pull request merged**